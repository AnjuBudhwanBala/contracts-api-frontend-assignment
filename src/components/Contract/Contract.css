.ContractItem {
  display: inline-block;
  width: 20%;
  vertical-align: top;
}
@media screen and (max-width: 1024px) {
  .ContractItem {
    width: 25%;
  }
}
@media screen and (max-width: 768px) {
  .ContractItem {
    width: 50%;
  }
}
@media screen and (max-width: 600px) {
  .ContractItem {
    width: 100%;
  }
}
.CompanyName {
  width: 100%;
  text-align: center;
}
.CompanyIcon {
  background: rgb(0, 132, 132);
  height: 50px;
  width: 50px;
  border-radius: 50%;
  color: #fff;
  position: relative;
  margin: 10px auto;
}
.CompanyIcon i {
  text-align: center;
  position: absolute;
  left: 50%;
  right: 0px;
  top: 50%;
  transform: translate(-50%, -50%);
}
.ContractDetails {
  padding: 30px;
  margin: 10px;
  box-shadow: 2px 2px 10px 1px #ccc;
  text-align: left;
  background: #fff;
  min-height: 360px;
  max-height: 400px;
  position: relative;
}

.ContractDetails ul {
  list-style: none;
  padding: 0;
  margin: 0;
}

.ContractDetails li {
  list-style: none;
  padding: 10px 0;
  border-bottom: 1px solid #ccc;
}
.ContractDetails li:first-child {
  color: rgb(0, 132, 132);
  text-align: center;
  font-weight: bold;
  margin-bottom: 20px;
  border: none;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.disabledText {
  color: #888;
  font-size: 12px;
  margin-top: 20px;
}

.ContractDetails .CustomButton {
  position: absolute;
  bottom: 20px;
  left: 50%;
  transform: translate(-50%);
}
