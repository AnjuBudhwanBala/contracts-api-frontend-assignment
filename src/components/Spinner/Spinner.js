import React from 'react';
import './Spinner.css';

const Spinner = ({ isLoading }) => {
  let spinner = null;
  if (isLoading) {
    spinner = (
      <div className="SpinnerOverlay">
        <div className="SpinnerContainer"></div>
      </div>
    );
  }

  return spinner;
};

export default Spinner;
