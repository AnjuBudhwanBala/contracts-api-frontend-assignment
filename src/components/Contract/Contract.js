import React, { useState } from 'react';
import 'font-awesome/css/font-awesome.min.css';
import './Contract.css';
import CustomButton from '../CustomButton/CustomButton';
import Modal from '../../components/Modal/Modal';
import CustomDateConverter from '../CustomDateConverter/CustomDateConverter';

const Contract = ({ contractInfo }) => {
  const [modalState, setModalState] = useState(false);
  const {
    company,
    negotiationRenewalDate,
    periodEnd,
    periodStart,
    scheduleForRenewal
  } = contractInfo;

  const openModalHandler = () => {
    setModalState(true);
  };
  const closeModalHandler = () => {
    setModalState(false);
  };

  return (
    <div className="ContractItem">
      <div className="ContractDetails">
        <div className="CompanyName">
          <div className="CompanyIcon">
            <i className="fa fa-industry" aria-hidden="true"></i>
          </div>
        </div>
        <ul>
          <li>
            <span>{company} </span>
          </li>
          <li>
            <strong>Start : </strong>
            <span>
              <CustomDateConverter date={periodStart} />
            </span>
          </li>
          <li>
            <strong>End : </strong>
            <span>
              <CustomDateConverter date={periodEnd} />
            </span>
          </li>
          {scheduleForRenewal ? (
            <li>
              <strong> Renewal Date : </strong>
              <span>
                <CustomDateConverter date={negotiationRenewalDate} />
              </span>
            </li>
          ) : (
            <span className="disabledText">Not schedule for renewal</span>
          )}
        </ul>
        <CustomButton
          click={openModalHandler}
          disabled={modalState}
          btnType="button"
        >
          Edit
        </CustomButton>
      </div>
      {modalState ? (
        <Modal
          contractInfo={contractInfo}
          closeModal={closeModalHandler}
          modalState={modalState}
        ></Modal>
      ) : null}
    </div>
  );
};

export default Contract;
