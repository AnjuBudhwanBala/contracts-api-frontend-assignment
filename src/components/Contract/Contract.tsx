import React, { useState } from 'react';
import 'font-awesome/css/font-awesome.min.css';
import './Contract.css';
import CustomButton from '../CustomButton/CustomButton';
import Modal from '../Modal/Modal';
import CustomDateConverter from '../CustomDateConverter/CustomDateConverter';
import Form from '../Form/Form';

export interface ContractObj {
  company: string;
  negotiationRenewalDate: string;
  periodEnd: string;
  periodStart: string;
  scheduleForRenewal: boolean;
  contractId?: string;
}

interface Props {
  contractInfo: ContractObj;
  refreshContracts: () => void;
}

export const Contract: React.FC<Props> = props => {
  const { contractInfo, refreshContracts } = props;

  const [isModalOpen, setIsModalOpen] = useState(false);
  const {
    company,
    negotiationRenewalDate,
    periodEnd,
    periodStart,
    scheduleForRenewal
  } = contractInfo;

  // to open Modal
  const openModalHandler = () => {
    setIsModalOpen(true);
  };
  //to close Modal
  const closeModalHandler = () => {
    setIsModalOpen(false);
  };
  //update contract after refresh
  const updateContract = () => {
    refreshContracts();
  };

  return (
    <div className="ContractItem">
      <div className="ContractDetails">
        <div className="CompanyName">
          <div className="CompanyIcon">
            <i className="fa fa-industry" aria-hidden="true"></i>
          </div>
        </div>
        <ul>
          <li>
            <span>{company}</span>
          </li>
          <li>
            <strong>Start : </strong>
            <span>
              <CustomDateConverter date={periodStart} />
            </span>
          </li>
          <li>
            <strong>End : </strong>
            <span>
              <CustomDateConverter date={periodEnd} />
            </span>
          </li>
          {scheduleForRenewal ? (
            <li>
              <strong> Renewal Date : </strong>
              <span>
                <CustomDateConverter date={negotiationRenewalDate} />
              </span>
            </li>
          ) : (
            <span
              style={{ color: '#888', fontSize: '12px', marginTop: '20px' }}
            >
              Not schedule for renewal
            </span>
          )}
        </ul>
        <CustomButton click={openModalHandler} btnType="button">
          Edit
        </CustomButton>
      </div>
      {isModalOpen ? (
        <Modal closeModal={closeModalHandler} modalState={isModalOpen}>
          <Form
            contractInfo={contractInfo}
            updateContract={updateContract}
            closeModal={closeModalHandler}
          ></Form>
        </Modal>
      ) : null}
    </div>
  );
};

export default Contract;
