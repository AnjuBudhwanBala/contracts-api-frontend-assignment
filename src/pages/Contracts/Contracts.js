import React, { useEffect, useState } from 'react';
import axios from '../../axiosUrl';
import Spinner from '../../components/Spinner/Spinner';
import Contract from '../../components/Contract/Contract';

export const Contracts = () => {
  const [contractsData, setContractsData] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);

  const fetchContracts = () => {
    axios
      .get('/contracts')
      .then(response => {
        setContractsData(response.data);
        setLoading(false);
      })
      .catch(error => {
        setLoading(false);
        setError(true);
      });
  };

  useEffect(() => {
    fetchContracts();
    return fetchContracts;
  }, []);

  const refreshContracts = () => {
    fetchContracts();
  };

  if (loading) {
    return <Spinner isLoading={loading} />;
  } else {
    if (error) {
      return <p id="error">Sorry we are unable to fetch Contracts</p>;
    } else {
      if (contractsData.length === 0) {
        return <p>You do not have any contracts</p>;
      } else {
        return contractsData.map(data => (
          <Contract
            key={data.contractId}
            contractInfo={data}
            refreshContracts={refreshContracts}
          ></Contract>
        ));
      }
    }
  }
};

export default Contracts;
