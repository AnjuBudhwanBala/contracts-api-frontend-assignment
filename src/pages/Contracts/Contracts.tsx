import React, { useEffect, useReducer } from 'react';
import axios from '../../axiosUrl';
import Spinner from '../../components/Spinner/Spinner';
import Contract from '../../components/Contract/Contract';
import { ContractObj } from '../../components/Contract/Contract';

interface Props {}

type State = {
  loading: boolean;
  error: boolean;
  contractsData: ContractObj[];
};

type Action =
  | { type: 'FETCH_SUCCESS'; payload: ContractObj[] }
  | { type: 'FETCH_ERROR' };

const initialState: State = {
  loading: true,
  error: false,
  contractsData: []
};

const reducer = (state: State, action: Action) => {
  switch (action.type) {
    case 'FETCH_SUCCESS':
      return {
        loading: false,
        error: false,
        contractsData: action.payload
      };
    case 'FETCH_ERROR':
      return {
        loading: false,
        error: true,
        contractsData: []
      };
    default: {
      return state;
    }
  }
};
export const Contracts = () => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const fetchContracts = () => {
    axios
      .get('/contracts')
      .then(response => {
        dispatch({ type: 'FETCH_SUCCESS', payload: response.data });
      })
      .catch(error => {
        dispatch({ type: 'FETCH_ERROR' });
      });
  };

  useEffect(() => {
    fetchContracts();
    return fetchContracts;
  }, []);

  const refreshContracts = () => {
    fetchContracts();
  };

  if (state.loading) {
    return (
      <>
        <Spinner isLoading={state.loading} />
      </>
    );
  } else {
    if (state.error) {
      return (
        <>
          <p id="error">Sorry we are unable to fetch Contracts</p>
        </>
      );
    } else {
      if (state.contractsData.length === 0) {
        return (
          <>
            <p>You do not have any contracts</p>
          </>
        );
      } else {
        return (
          <>
            {state.contractsData.map(data => (
              <Contract
                key={data.contractId}
                contractInfo={data}
                refreshContracts={refreshContracts}
              ></Contract>
            ))}
          </>
        );
      }
    }
  }
};

export default Contracts;
