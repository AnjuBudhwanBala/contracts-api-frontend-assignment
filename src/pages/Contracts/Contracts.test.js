import React from 'react';
import { shallow } from 'enzyme';
import Spinner from '../../components/Spinner/Spinner';
import Contract from '../../components/Contract/Contract';
import { Contracts } from './Contracts';

describe('Contracts Component', () => {
  let wrapper;

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render Contracts component', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
