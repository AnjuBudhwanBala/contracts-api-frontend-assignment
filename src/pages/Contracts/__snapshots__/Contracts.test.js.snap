// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contracts Component should render Contracts component 1`] = `undefined`;
